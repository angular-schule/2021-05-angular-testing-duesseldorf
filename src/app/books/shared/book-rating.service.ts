import { Injectable } from '@angular/core';
import { Book } from './book';

@Injectable({
  providedIn: 'root'
})
export class BookRatingService {

  constructor() { }

  rateUp(book: Book): Book {
    return {
      ...book,
      rating: Math.min(5, book.rating + 1)
    };
  }

  rateDown(book: Book): Book {
    return {
      ...book,
      rating: Math.max(1, book.rating - 1)
    };
  }
}
