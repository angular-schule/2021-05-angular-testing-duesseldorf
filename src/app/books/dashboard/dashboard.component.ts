import { Component, OnInit } from '@angular/core';
import { of } from 'rxjs';

import { Book } from '../shared/book';
import { BookRatingService } from '../shared/book-rating.service';
import { BookStoreService } from '../shared/book-store.service';

@Component({
  selector: 'br-dashboard',
  templateUrl: './dashboard.component.html',
  styleUrls: ['./dashboard.component.scss']
})
export class DashboardComponent implements OnInit {

  books: Book[];

  constructor(
    // AKTIVIEREN für HTTP
    // private bs: BookStoreService,
    private rs: BookRatingService
    ) { }

    ngOnInit() {
    // AKTIVIEREN für HTTP
    // this.bs.getAll().subscribe(books => this.books = books);
    this.books = [
      {
        isbn: '111',
        title: 'Angular',
        description: 'Grundlagen und Best Practices',
        authors: ['Ferdinand Malcher', 'Johannes Hoppe'],
        rating: 5,
        firstThumbnailUrl: null
      },
      {
        isbn: '222',
        title: 'React',
        description: 'Das andere Framework',
        authors: ['Max Mustermann', 'Erika Musterfrau'],
        rating: 3,
        firstThumbnailUrl: null
      },
    ];
  }

  rateUp(book: Book) {
    const ratedBook = this.rs.rateUp(book);
    this.updateList(ratedBook);
  }

  rateDown(book: Book) {
    const ratedBook = this.rs.rateDown(book);
    this.updateList(ratedBook);
  }

  updateList(ratedBook: Book) {
    this.books = this.books
      .map(b => b.isbn === ratedBook.isbn ? ratedBook : b);
  }

  trackBook(index: number, item: Book) {
    return item.isbn;
  }
}
