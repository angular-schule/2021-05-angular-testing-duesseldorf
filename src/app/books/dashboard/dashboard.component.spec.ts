import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { of } from 'rxjs';
import { Book } from '../shared/book';
import { BookRatingService } from '../shared/book-rating.service';
import { BookStoreService } from '../shared/book-store.service';

import { DashboardComponent } from './dashboard.component';

describe('DashboardComponent', () => {
  let component: DashboardComponent;
  let fixture: ComponentFixture<DashboardComponent>;
  let book: Book;
  let ratingMock: Partial<BookRatingService>;
  let bsMock: Partial<BookStoreService>;

  beforeEach(async () => {
    ratingMock = {
      rateUp: jest.fn(b => b),
    };

    bsMock = {
      getAll: jest.fn().mockReturnValue(of([])),
    };

    book = {
      isbn: '',
      title: '',
      description: '',
      rating: 3,
      authors: [],
      firstThumbnailUrl: null,
    };

    await TestBed.configureTestingModule({
      declarations: [DashboardComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        // BRS ersetzen: Wenn BRS angefordert wird, wird ratingMock ausgeliefert
        { provide: BookRatingService, useValue: ratingMock },
        { provide: BookStoreService, useValue: bsMock },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(DashboardComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should call the service method rateUp() for rateUp()', () => {
    // Act
    component.rateUp(book); // wird sonst durch Event aus dem Template aufgerufen

    // Assert (Verhalten)
    expect(ratingMock.rateUp).toHaveBeenCalledTimes(1);
    expect(ratingMock.rateUp).toHaveBeenCalledWith(book);
  });

  it('should replace the updated book', () => {
    component.books = [
      { isbn: '111', title: 'A' },
      { isbn: '222', title: 'B' },
      { isbn: '333', title: 'C' },
      { ...book, isbn: '444' },
    ] as Book[]; // type assertion!

    const changedBook = { isbn: '222', title: 'CHANGED!' } as Book;

    // Act
    component.updateList(changedBook);

    // Assert
    expect(component.books).toMatchSnapshot();
  });
});
