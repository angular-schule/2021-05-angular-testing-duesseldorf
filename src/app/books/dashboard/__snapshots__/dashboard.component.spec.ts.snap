// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardComponent should replace the updated book 1`] = `
Array [
  Object {
    "isbn": "111",
    "title": "A",
  },
  Object {
    "isbn": "222",
    "title": "CHANGED!",
  },
  Object {
    "isbn": "333",
    "title": "C",
  },
  Object {
    "authors": Array [],
    "description": "",
    "firstThumbnailUrl": null,
    "isbn": "444",
    "rating": 3,
    "title": "",
  },
]
`;
